src/main/java/org/lantern/PublicIpInfoHandler.java

```diff
@@ -46,72 +42,66 @@ public PublicIpInfoHandler(final Model model, final Censored censored,
+            } catch (final ConnectException e) {
+                log.warn("Could not get public IP?", e);
+            }
+        } else {
+            log.debug("Not connected to the internet, not determining public ip");
+        }
+    }
+
     /**
-     * We set the IP address on a proxy connection because we use the proxy
-     * itself to determine the IP address. This helps to minimize Lantern's
-     * network footprint.
+     * Determines the public ip address using PublicIpAddress.
      * 
-     * @throws ConnectException If there was an error fetching the public IP. 
+     * @throws ConnectException
+     *             If there was an error fetching the public IP.
      */
     private void determinePublicIp() throws ConnectException {
-        if (!this.recheck.get()) {
-            log.debug("Not rechecking for public IP");
-            return;
+        InetAddress address = null;
+        for (int i = 0; i < 10; i++) {
+            // Try a few times to get the public ip address
+            // We do this because for some reason, if we've just restored
+            // connectivity and gotten a connected event, connecting to the
+            // geo service to do an IP lookup sometimes fails.
+            try {
+                // Back off a little on each try
+                Thread.sleep(i * i * 100);
+            } catch (InterruptedException ie) {
+                // ignore
+            }
+            address = new PublicIpAddress().getPublicIpAddress();
+            if (address != null) {
+                break;
+            }
         }
-        
-        final InetAddress address = new PublicIpAddress().getPublicIpAddress();
-        this.model.getConnectivity().setIp(address != null ? address.getHostAddress() : null);
+
+        this.model.getConnectivity().setIp(
+                address != null ? address.getHostAddress() : null);
         if (address == null) {
             throw new ConnectException("Could not determine public IP");
         }
-        
-        this.recheck.set(false);
+
         handleCensored(address);
         handleGeoIp(address);
-        
+
         // Post PublicIpEvent so that downstream services like xmpp,
         // FriendsHandler, StatsManager and Loggly can start.
         Events.asyncEventBus().post(new PublicIpEvent());
     }
-    
-    @Subscribe
-    public void onConnectivityChanged(final ConnectivityChangedEvent event) {
-        // Make sure we re-check your public IP after any change in 
-        // connectivity.
-        this.recheck.set(true);
-    }
-    
-    /**
-     * We perform the public IP lookup on a proxy connection event because
-     * we need the proxy in order to perform the lookup. We need this both
-     * at startup as well as every time we re-connect to a proxy after
-     * potentially losing proxy connectivity.
-     * 
-     * @param pce The connection event.
-     */
-    @Subscribe
-    public void onProxyConnectionEvent(
```

Comment:
This dependency from PublicIp to ProxyConnection was really outdated, since we've been using a host-spoofed Get to geoip for a while now.  This realization simplified PublicIpInfoHandler quite a bit, since we can now just check every time that connectivity is established/restored.